File: tt-train/sources/ttml/ops/sink_attention.cpp
// SPDX-FileCopyrightText: (c) 2026 Tenstorrent USA, Inc.
//
// SPDX-License-Identifier: Apache-2.0

#include "sink_attention.hpp"

#include <cmath>
#include <core/ttnn_all_includes.hpp>

#include "ops/binary_ops.hpp"
#include "ops/matmul_op.hpp"
#include "ops/unary_ops.hpp"

namespace ttml::ops {

autograd::TensorPtr shared_kv_mqa_attention(
    const autograd::TensorPtr& q,
    const autograd::TensorPtr& kv,
    const autograd::TensorPtr& sink_logits,
    const std::optional<autograd::TensorPtr>& keep_mask) {
    const auto q_shape = q->get_value().logical_shape().to_array_4D();
    const uint32_t batch = q_shape[0];
    const uint32_t heads = q_shape[1];
    const uint32_t seq = q_shape[2];
    const uint32_t head_dim = q_shape[3];
    const uint32_t groups = kv->get_value().logical_shape().to_array_4D()[2];
    const float scale = 1.0F / std::sqrt(static_cast<float>(head_dim));

    // Share the compressed KV entries across all query heads (MQA).
    auto kv_bc = ops::broadcast_to(kv, ttnn::Shape({batch, heads, groups, head_dim}));  // [B, H, G, c]

    // Scaled logits L = (q . kv^T) / sqrt(c).
    auto logits = ops::mul(ops::matmul_op(q, kv_bc, /* transpose_a */ false, /* transpose_b */ true), scale);

    // Apply the selection mask as an additive bias (0 for kept, large negative for
    // dropped). The mask carries no gradient, so it is folded in as a constant.
    if (keep_mask.has_value()) {
        constexpr float kBig = 1.0e4F;
        auto m = keep_mask.value()->get_value();                    // [B, 1, S, G]
        auto bias = ttnn::multiply(ttnn::subtract(m, 1.0F), kBig);  // 0 / -kBig
        bias = ttnn::repeat(bias, ttnn::Shape({1, heads, 1, 1}));   // -> [B, H, S, G]
        logits = ops::add(logits, bias);
    }

    // Sink-augmented softmax over the G compressed blocks. RMSNorm upstream keeps
    // the logits bounded, so exp is taken directly (no max-subtraction).
    auto numerator = ops::exp(logits);                                    // [B, H, S, G]
    auto denom = ops::sum(numerator, /* dim */ -1, /* keep_dim */ true);  // [B, H, S, 1]
    auto sink = ops::broadcast_to(ops::exp(sink_logits), ttnn::Shape({batch, heads, seq, 1U}));
    denom = ops::add(denom, sink);  // + exp(z'_h)
    auto weights = ops::div(numerator, ops::broadcast_to(denom, ttnn::Shape({batch, heads, seq, groups})));

    // Weighted sum of the (shared) value entries.
    return ops::matmul_op(weights, kv_bc, /* transpose_a */ false, /* transpose_b */ false);  // [B, H, S, c]
}

}  // namespace ttml::ops

File: models/demos/deepseek_v4_inference/tt/sink_attention.py
# SPDX-FileCopyrightText: (c) 2026 Tenstorrent USA, Inc.
#
# SPDX-License-Identifier: Apache-2.0

"""TT-NN shared-KV MQA + attention sink (roadmap step 2.4, the math core).

Validated against `reference_model.masked_sink_attention`. This is the *math* the
production top-k DRAM-gather decode kernel realizes: a query attends over the kept
compressed/window KV (here selected by a 0/1 keep mask rather than physically
gathered), with a per-head attention sink added to the softmax denominator so
fully-masked rows are numerically safe (-> 0).

    L = (q . kv) / sqrt(hd) + (keep-1)*BIG          # masked logits
    w = exp(L - m) / ( sum exp(L - m) + exp(sink - m) )     (m = rowmax, for stability)
    o = w @ kv

The production realization (next sub-step) replaces the dense [S, K] mask with a
gather of only the top-k blocks from DRAM (extend deepseek_v3_b1 flash_mla +
kv_cache_update). Mirrors the tt-train C++ shared_kv_mqa_attention.
"""

from __future__ import annotations

import ttnn

_BIG = 1.0e9


def sink_attention(q: "ttnn.Tensor", keys: "ttnn.Tensor", keep: "ttnn.Tensor",
                   sink: "ttnn.Tensor", scale: float) -> "ttnn.Tensor":
    """q: [b,H,S,hd], keys (=values): [b,1,K,hd], keep: [b,1,S,K] (0/1), sink: [1,H,1,1].
    Returns [b,H,S,hd]."""
    b, heads, s, hd = q.shape
    keys_h = ttnn.repeat(keys, ttnn.Shape((1, heads, 1, 1)))             # [b, H, K, hd] (MQA share)
    logits = ttnn.multiply(ttnn.matmul(q, ttnn.transpose(keys_h, -2, -1)), scale)  # [b, H, S, K]

    bias = ttnn.multiply(ttnn.subtract(keep, 1.0), _BIG)                # 0 (keep) / -BIG (drop), [b,1,S,K]
    logits = ttnn.add(logits, bias)                                    # broadcast over H

    m = ttnn.max(logits, dim=-1, keepdim=True)                         # [b, H, S, 1] row max (stability)
    e = ttnn.exp(ttnn.subtract(logits, m))
    sink_term = ttnn.exp(ttnn.subtract(sink, m))                      # exp(sink_h - m), broadcast over S,K->1
    denom = ttnn.add(ttnn.sum(e, dim=-1, keepdim=True), sink_term)
    w = ttnn.div(e, denom)
    return ttnn.matmul(w, keys_h)                                      # [b, H, S, hd]
